inflate.c

```diff
@@ -1149,17 +1150,33 @@ int ZEXPORT inflate(z_stream *strm, int flush) {
                 }
                 if (copy > state->length)
                     copy = state->length;
+                if (copy > left)
+                    copy = left;
+                left -= copy;
+                state->length -= copy;
+                if (copy >= sizeof(long)) {
+                    put = chunk_memcpy(put, from, copy);
+                } else {
+                    do {
+                        *put++ = *from++;
+                    } while (--copy);
+                }
             } else {                             /* copy from output */
-                from = put - state->offset;
+                unsigned offset = state->offset;
+                from = put - offset;
                 copy = state->length;
+                if (copy > left)
+                    copy = left;
+                left -= copy;
+                state->length -= copy;
+                if (copy >= sizeof(long)) {
+                    put = chunk_memset(put, from, offset, copy);
```

Comment:
This might be a problem.  `chunk_memset()` assumes that there's room to write at least two times the chunks size to the output to implement the unroll.  How much it needs is variable, but it should always be less than `2*sizeof(...)`.

That said, if it doesn't need to unroll (if the copies don't overlap) then we could test `left >= sizeof(...)` instead of `copy >= sizeof(...)`; which allows it to paint with a fatter brush.